lib/oban/queue/basic_engine.ex

```diff
@@ -141,16 +141,24 @@ defmodule Oban.Queue.BasicEngine do
 
   @impl true
   def cancel_job(%Config{} = conf, %Job{id: id}) do
+    query = where(Job, [j], j.id == ^id)
+
+    cancel_all_jobs(conf, query)
+
+    :ok
+  end
+
+  @impl true
+  def cancel_all_jobs(%Config{} = conf, queryable) do
     query =
-      Job
-      |> where([j], j.id == ^id)
+      queryable
       |> where([j], j.state not in ["completed", "discarded", "cancelled"])
+      |> update([j], set: [state: "cancelled", cancelled_at: ^utc_now()])
+      |> select([:id])
```

Comment:
We can avoid the extra map by selecting the id directly:

```suggestion
      |> select([j], j.id)
```